docs/design.md

```diff
@@ -0,0 +1,326 @@
+to the root CA will be signing the certificates for users and hosts. The host and user will
+need to have a copy of this root certificate in order to verify each other.
+
+![certificate-hierarchy](./certificate_hierarchy.png)
+
+### Cipher Suites
+
+#### TLS 1.2
+
+Following the guidance from these RFCs:
+
+[Recommendations for Secure Use of Transport Layer Security and Datagram Transport Layer Security (DTLS)](https://www.ietf.org/archive/id/draft-ietf-uta-rfc7525bis-03.html)
+
+[Hypertext Transfer Protocol Version 2 (HTTP/2) Appendix A](https://datatracker.ietf.org/doc/html/rfc7540#appendix-A)
+
+We will only accept cipher suites that match the following criteria:
+* Key exchange algorithm supports perfect forward secrecy
+* No RC4 cipher suites
+* No cipher suites including algorithms offering less than 128 bits of security
+* No cipher suites which use static RSA i.e cipher suites starting with TLS_RSA_WITH_*
+* Cipher Suite is not blacklisted by HTTP/2
+
+This leaves us with the following cipher suites supported by the Go TLS package https://pkg.go.dev/crypto/tls :
+
+TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+
+#### TLS 1.3
+
+TLS 1.3 offers enhanced security over its predecessor, all its supported cipher suites can be negotiated.
+
+### Authorisation
+
+For authorisation, a simple all-or-nothing mechanism will be used. The server can be started with a list of OS users that are allowed to make requests.
+
+X.509 certificates issued to users will have information about the user name encoded in them in the Common Name subject field. The server will then use this information to decide if a requester has been given permission to make requests.
+
+For example, using the snippet below the users `hashi`, `mozart` and `simba` would be able to make requests to the GRPC API.
+
+```
+./worker --allowedUsers=hashi,mozart,simba
+```
+
+## Worker Library
+
+The service will include a worker library written in Go to manage and interact with Linux and Linux processes.
+Under the hood, this is what the GRPC server will use to fulfill requests.
+
+A sketch of what this library could look like is below:
+
+```go
+// Worker is a Linux process manager
+// It keeps a registry of the processes that it
+// been requested to manage in memory.
+type Worker struct {
+	defaultresourceLimits ResourceLimits
+	execRegistry          map[ID]*ProcessHandle
+}
+
+// ID is the internal ID that the worker uses to identify an Exec
+// Under the hood this might just be a UUID
+type ID string
+
+type ResourceLimits struct {
+	MaxMemoryBytes int64
+	// ...
+}
+
+// ProcessRequest represents a request to execute a Linux process in the worker
+type ProcessRequest struct {
+	// Command is the command executed to be executed together with its arguments
+	Command string
```

Comment:
Suggestion: use a slice / repeated field for args, parsing command-line strings is not trivial when you consider quoting.

Reply:
+1 on this suggestion from me